<a href="https://colab.research.google.com/github/shinhs0920/Engineering-computing-application/blob/master/LSTM_two_layer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
try:

    %tensorflow_version 2.x  # %tensorflow_version only exists in Colab

except Exception:

    pass

`%tensorflow_version` only switches the major version: `1.x` or `2.x`.
You set: `2.x  # %tensorflow_version only exists in Colab`. This will be interpreted as: `2.x`.


TensorFlow is already loaded. Please restart the runtime to change versions.


In [5]:
import tensorflow as tf
tf.__version__

'2.0.0'

---
> ### 텐서플로로 LSTM 순환 신경망 만들기

먼저 LSTM을 한층만 설계하여 진행한 후 LSTM을 한 층 더 추가하여 정확도를 비교하겠습니다.

In [6]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LSTM

model_lstm = Sequential()

model_lstm.add(Embedding(1000,32))
model_lstm.add(LSTM(8))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 32)          32000     
_________________________________________________________________
lstm_1 (LSTM)                (None, 8)                 1312      
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 9         
Total params: 33,321
Trainable params: 33,321
Non-trainable params: 0
_________________________________________________________________


LSTM을 한 층으로 모델을 설정하였습니다.

10번의 에포크 동안 Adam 최적화 알고리즘을 사용하여 모델을 훈련했습니다.

In [7]:
from tensorflow.keras.datasets import imdb
(x_train_all, y_train_all), (x_test, y_test)=imdb.load_data(skip_top=20, num_words=1000)
# 앞과 마찬가지로 단어 1000개

for i in range(len(x_train_all)):
  x_train_all[i]=[w for w in x_train_all[i] if w>2]

import numpy as np
np.random.seed(42)
random_index=np.random.permutation(25000)

x_train=x_train_all[random_index[:20000]]
y_train=y_train_all[random_index[:20000]]
x_val=x_train_all[random_index[20000:]]
y_val=y_train_all[random_index[20000:]]



17465344/17464789 [==============================] - 0s 0us/step


In [0]:
from tensorflow.keras.preprocessing import sequence

maxlen=100
x_train_seq = sequence.pad_sequences(x_train, maxlen=maxlen)
x_val_seq = sequence.pad_sequences(x_val, maxlen=maxlen)

In [9]:
model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
history=model_lstm.fit(x_train_seq, y_train, epochs=10, batch_size=32, validation_data=(x_val_seq, y_val))

Train on 20000 samples, validate on 5000 samples
Epoch 1/10
20000/20000 [==============================] - 28s 1ms/sample - loss: 0.4473 - accuracy: 0.8023 - val_loss: 0.3727 - val_accuracy: 0.8428
Epoch 2/10
20000/20000 [==============================] - 24s 1ms/sample - loss: 0.3321 - accuracy: 0.8595 - val_loss: 0.3647 - val_accuracy: 0.8460
Epoch 3/10
20000/20000 [==============================] - 24s 1ms/sample - loss: 0.3100 - accuracy: 0.8706 - val_loss: 0.3545 - val_accuracy: 0.8456
Epoch 4/10
20000/20000 [==============================] - 24s 1ms/sample - loss: 0.2934 - accuracy: 0.8774 - val_loss: 0.3621 - val_accuracy: 0.8456
Epoch 5/10
20000/20000 [==============================] - 23s 1ms/sample - loss: 0.2798 - accuracy: 0.8834 - val_loss: 0.3704 - val_accuracy: 0.8394
Epoch 6/10
20000/20000 [==============================] - 23s 1ms/sample - loss: 0.2652 - accuracy: 0.8888 - val_loss: 0.3866 - val_accuracy: 0.8436
Epoch 7/10
20000/20000 [==============================] -

##### 정확도를 확인합니다.

In [12]:
loss, accuracy = model_lstm.evaluate(x_val_seq, y_val, verbose=0)
print(accuracy)

0.8332


검증 세트의 정확도가 83.32% 입니다.

###다음은 LSTM을 한 층 더 추가하여 두개의 층으로 실행하겠습니다.

In [24]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LSTM

model_lstm = Sequential()
model_lstm.add(Embedding(1000,32))
model_lstm.add(LSTM(8, return_sequences=True))
model_lstm.add(LSTM(4,return_sequences=False))
model_lstm.add(Dense(1, activation='sigmoid'))

model_lstm.summary()

Model: "sequential_12"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_12 (Embedding)     (None, None, 32)          32000     
_________________________________________________________________
lstm_11 (LSTM)               (None, None, 8)           1312      
_________________________________________________________________
lstm_12 (LSTM)               (None, 4)                 208       
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 5         
Total params: 33,525
Trainable params: 33,525
Non-trainable params: 0
_________________________________________________________________


return_sequences를 이용하여 LSTM을 한 층 더 추가하여 총 2층의 LSTM을 설계하였습니다.

In [0]:
from tensorflow.keras.datasets import imdb
(x_train_all, y_train_all), (x_test, y_test)=imdb.load_data(skip_top=20, num_words=1000)
# 앞과 마찬가지로 단어 1000개

for i in range(len(x_train_all)):
  x_train_all[i]=[w for w in x_train_all[i] if w>2]

import numpy as np
np.random.seed(42)
random_index=np.random.permutation(25000)

x_train=x_train_all[random_index[:20000]]
y_train=y_train_all[random_index[:20000]]
x_val=x_train_all[random_index[20000:]]
y_val=y_train_all[random_index[20000:]]


In [0]:
from tensorflow.keras.preprocessing import sequence

maxlen=100
x_train_seq = sequence.pad_sequences(x_train, maxlen=maxlen)
x_val_seq = sequence.pad_sequences(x_val, maxlen=maxlen)

In [28]:
model_lstm.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

history=model_lstm.fit(x_train_seq, y_train, epochs=10, batch_size=32, validation_data=(x_val_seq, y_val))

Train on 20000 samples, validate on 5000 samples
Epoch 1/10
20000/20000 [==============================] - 51s 3ms/sample - loss: 0.4868 - accuracy: 0.7798 - val_loss: 0.3934 - val_accuracy: 0.8356
Epoch 2/10
20000/20000 [==============================] - 45s 2ms/sample - loss: 0.3511 - accuracy: 0.8553 - val_loss: 0.3593 - val_accuracy: 0.8448
Epoch 3/10
20000/20000 [==============================] - 45s 2ms/sample - loss: 0.3133 - accuracy: 0.8715 - val_loss: 0.3710 - val_accuracy: 0.8348
Epoch 4/10
20000/20000 [==============================] - 46s 2ms/sample - loss: 0.2914 - accuracy: 0.8837 - val_loss: 0.3650 - val_accuracy: 0.8430
Epoch 5/10
20000/20000 [==============================] - 46s 2ms/sample - loss: 0.2760 - accuracy: 0.8881 - val_loss: 0.3693 - val_accuracy: 0.8358
Epoch 6/10
20000/20000 [==============================] - 46s 2ms/sample - loss: 0.2619 - accuracy: 0.8935 - val_loss: 0.3803 - val_accuracy: 0.8402
Epoch 7/10
20000/20000 [==============================] -

##### 정확도를 확인합니다.

In [31]:
loss, accuracy = model_lstm.evaluate(x_val_seq, y_val, verbose=0)
print(accuracy)

0.8374


LSTM 층을 하나 더 추가하였더니 83.32%에서 83.74%로 정확도가 조금 상승한 것을 볼 수 있었습니다.